modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/HttpHeadersValidator.java

```diff
@@ -0,0 +1,178 @@
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.http.netty4.Netty4HttpRequest.getHttpHeadersAsMap;
+import static org.elasticsearch.http.netty4.Netty4HttpRequest.translateRequestMethod;
+
+/**
+ * Provides utilities used to hook into the netty pipeline and run a validation check on each HTTP request's headers.
+ * See also {@link Netty4HttpHeaderValidator}.
+ */
+public class HttpHeadersValidator {
+
+    /**
+     * Trivial {@link HttpHeadersValidator} implementation, to be used in tests, that successfully validates
+     * any and all HTTP request headers.
+     */
+    public static final HttpHeadersValidator VALIDATE_EVERYTHING_VALIDATOR = new HttpHeadersValidator(
+        (httpPreRequest, channel, listener) -> listener.onResponse(ValidatableHttpHeaders.OK)
+    );
+
+    /**
+     * An async HTTP headers validator function that receives as arguments part of the incoming HTTP request
+     * (except the body contents, see {@link HttpPreRequest}), as well as the netty channel that the request is
+     * being received over, and must then call the {@code ActionListener#onResponse} method on the listener parameter
+     * in case the validation is to be considered successful, or otherwise call {@code ActionListener#onFailure}.
+     */
+    private final TriConsumer<HttpPreRequest, Channel, ActionListener<ValidationResultContext>> validator;
+
+    public HttpHeadersValidator(TriConsumer<HttpPreRequest, Channel, ActionListener<ValidationResultContext>> validator) {
+        this.validator = validator;
+    }
+
+    /**
+     * Supplies a netty {@code ChannelInboundHandler} that runs the {@link #validator} on the HTTP request headers.
+     * The HTTP headers of the to be validated {@link HttpRequest} must be wrapped by the special {@link ValidatableHttpHeaders},
+     * see {@link #wrapAsValidatableMessage(HttpMessage)}.
+     */
+    public Netty4HttpHeaderValidator getValidatorInboundHandler() {
+        return new Netty4HttpHeaderValidator((httpRequest, channel, listener) -> {
+            if (httpRequest.headers() instanceof ValidatableHttpHeaders validatableHttpHeaders) {
+                // make sure validation only runs on properly wrapped "validatable" headers implementation
+                validator.apply(asHttpPreRequest(httpRequest), channel, ActionListener.wrap(validationResultContext -> {
+                    validatableHttpHeaders.markAsSuccessfullyValidated(validationResultContext);
+                    // a successful validation needs to signal to the {@link Netty4HttpHeaderValidator} to resume
+                    // forwarding the request beyond the headers part
+                    listener.onResponse(null);
+                }, e -> listener.onFailure(new HttpHeadersValidationException(e))));
+            } else {
+                listener.onFailure(new IllegalStateException("Expected a validatable request for validation"));
+            }
+        });
+    }
```

Comment:
The overarching design is that in "netty land" , authentication doesn't exist. It's only validation and validators. It is the Security plugin that hooks into the "validation" and implements authentication.